dev/system-test/tracing.ts

```diff
@@ -0,0 +1,656 @@
+  let gcpTraceExporter: TraceExporter;
+
+  const spanIdToChildrenSpanIds = new Map<string, string[]>();
+  const spanIdToSpanData = new Map<string, ReadableSpan>();
+  let rootSpanIds: string[] = [];
+
+  function afterEachTest(config: TestConfig): Promise<void> {
+    spanIdToChildrenSpanIds.clear();
+    spanIdToSpanData.clear();
+    rootSpanIds = [];
+
+    return verifyInstance(firestore);
+  }
+
+  function getOpenTelemetryOptions(
+    config: TestConfig,
+    tracerProvider: TracerProvider
+  ): FirestoreOpenTelemetryOptions {
+    const options: FirestoreOpenTelemetryOptions = {
+      enableTracing: true,
+      tracerProvider: undefined,
+    };
+
+    // If we are *not* using a global OpenTelemetry instance, a TracerProvider
+    // must be passed to the Firestore SDK.
+    if (!config.globalOpenTelemetry) {
+      options.tracerProvider = tracerProvider;
+    }
+
+    return options;
+  }
+
+  function beforeEachTest(config: TestConfig) {
+    // Remove the global tracer provider in case anything was registered
+    // in order to avoid duplicate global tracers.
+    trace.disable();
+    context.disable();
+
+    // Set up a context manager.
+    const contextManager = new AsyncLocalStorageContextManager();
+    contextManager.enable();
+    context.setGlobalContextManager(contextManager);
+
+    console.log('in beforeEachTest');
+    // Create a new tracer and span processor for each test to make sure there
+    // are no overlaps when reading the results.
+    tracerProvider = new NodeTracerProvider({
+      sampler: new AlwaysOnSampler(),
+    });
+
+    inMemorySpanExporter = new InMemorySpanExporter();
+    consoleSpanExporter = new ConsoleSpanExporter();
+    gcpTraceExporter = new TraceExporter();
+
+    // Always add the console exporter for local debugging.
+    tracerProvider.addSpanProcessor(
+      new BatchSpanProcessor(consoleSpanExporter)
+    );
+
+    if (config.e2e) {
+      tracerProvider.addSpanProcessor(new BatchSpanProcessor(gcpTraceExporter));
+    } else {
+      tracerProvider.addSpanProcessor(
+        new BatchSpanProcessor(inMemorySpanExporter)
+      );
+    }
+
+    if (config.globalOpenTelemetry) {
+      trace.setGlobalTracerProvider(tracerProvider);
+    }
+
+    const settings: Settings = {
+      preferRest: config.preferRest,
+      openTelemetryOptions: getOpenTelemetryOptions(config, tracerProvider),
+    };
+    if (process.env.FIRESTORE_NAMED_DATABASE) {
+      settings.databaseId = process.env.FIRESTORE_NAMED_DATABASE;
+    }
+    if (!settings.projectId && process.env.PROJECT_ID) {
+      settings.projectId = process.env.PROJECT_ID;
+    }
+    if (!settings.databaseId && process.env.DATABASE_ID) {
+      settings.databaseId = process.env.DATABASE_ID;
+    }
+
+    firestore = new Firestore(settings);
+  }
+
+  // Returns true on success, and false otherwise.
+  async function waitForCompletedInMemorySpans(): Promise<boolean> {
+    await tracerProvider.forceFlush();
```

Comment:
does `forceFlush` return false if it fails?

Reply:
no, it's a `Promise<void>`. If the flush is not successful, we likely won't get the correct number of spans and the test will fail.